### PR TITLE
[RFC] Support full bf16 training

### PR DESCRIPTION
**Pull request by ebsmothers**
This PR adds support for full bfloat16 training. In SFT it is pretty common to store everything in bfloat16 to save memory, with select tensors (logits, RoPE buffers and activations) maintained in a higher precision to preserve numerical accuracy. Separately I think having this supported more generally would be useful for faster iteration -- e.g. it allows me to run Llama3 70B on a single node of H100s, which otherwise is not possible with the default config.

Assuming this is generally useful, would like feedback on:

1) Acceptable loss convergence: in the first 100 steps on Llama3 8B full bf16 training goes from 12.25 -> 8, as opposed to 12.25 -> 7 with fp32 training. Is this a concern? (As mentioned, for SFT this is less of an issue; happy to validate that statement if that's helpful.)
2) Interaction with mixed precision training -- where is the right place to validate that these are not both set at once?
3) Where to put the `set_default_dtype` API

### CODE REVIEW

File: torchtitan/models/llama3/model/model.py [hunk 1]
```diff
@@ -421,5 +421,5 @@ def forward(
             h = layer(h, self.freqs_cis)
 
         h = self.norm(h) if self.norm else h
-        output = self.output(h) if self.output else h
+        output = self.output(h).float() if self.output else h
```
Comment (fegin) [inline torchtitan/models/llama3/model/model.py]: If we set the training dtype during the training initialization, why not also do the output conversion in the trainer (train loop)?

File: torchtitan/config/job_config.py [hunk 2]
```diff
@@ -204,6 +204,13 @@ class Training:
...
     """
 
+    dtype: Literal["bfloat16", "float32"] = "float32"
+    """
+    torch dtype for training. In contrast to mixed precision training, setting training_dtype=bfloat16 will
+    put all parameters, gradients, and optimizer states in bfloat16, without an extra copy of fp32 weights.
+    In the case of full bf16 training, RoPE calculations and logits will still be in fp32.
+    """
+
     mixed_precision_param: Literal["bfloat16", "float32"] = "bfloat16"
```
Comment (fegin) [inline torchtitan/config/job_config.py:214]: What if `mixed_precision_param` is `float32` but dtype is `bfloat16`? There should be a check?

Comment (ebsmothers) [reply torchtitan/config/job_config.py:214]: Yeah agreed. Do we want to do this somewhere in train.py? Lmk if you think there's a better place

File: torchtitan/models/llama3/model/model.py (hunk 1 above)
Comment (tianyu-l) [reply torchtitan/models/llama3/model/model.py]: This is already in the loss function https://github.com/pytorch/torchtitan/blob/main/torchtitan/components/loss.py#L21

Also see https://github.com/pytorch/torchtitan/pull/642

File: torchtitan/config/job_config.py (hunk 2 above)
Comment (tianyu-l) [reply torchtitan/config/job_config.py:214]: `mixed_precision_param` is coming from FSDP2. I think if FSDP2 can work with that, it's users responsibility to config them properly.

Comment (fegin) [reply torchtitan/config/job_config.py:214]: We also make it work with DDP/single device: https://github.com/pytorch/torchtitan/pull/1303. I think a warning is at least required.

File: torchtitan/models/llama3/model/model.py (hunk 1 above)
Comment (ebsmothers) [reply torchtitan/models/llama3/model/model.py]: Thanks, just removed

File: torchtitan/config/job_config.py (hunk 2 above)
Comment (ebsmothers) [reply torchtitan/config/job_config.py:214]: Sounds good. In that case I will leave this as is

Comment (tianyu-l) [reply torchtitan/config/job_config.py:214]: @fegin 
`autocast` is not well supported in torchtitan anyways. I'm not sure if it is still maintained. See other issue like https://github.com/pytorch/torchtitan/issues/1525

But sure, having a warning sounds good.

**Comment by fegin** [comment]
One last thing in my mind is that the `set_default_dtype()` definition should be moved to `torchtitan/model`.

**Comment by tianyu-l** [comment]
@fegin 
> One last thing in my mind is that the `set_default_dtype()` definition should be moved to `torchtitan/model`

IIUC it is a general context manager, very similar to `with torch.device()`. Curious why you think it should be moved to `model` folder?

**Comment by fegin** [comment]
> IIUC it is a general context manager, very similar to with torch.device(). Curious why you think it should be moved to model folder?

Unlike `torch.device()` which allocates the same model on different devices, this context changes the entire model, the model meaning is technically different. That's my motivation. But I don't have a strong opinion on this. It is also reasonable to put under tool folder.

**Comment by tianyu-l** [comment]
Oh I see what you mean. I think the function itself can be used for more than model definition, so I'd still prefer it being in a util folder. Maybe let's merge it as is if you don't have strong opinion.

**Comment by hann-wang** [comment]
I don't think keeping optimizer states in BF16 is a good idea.

Generally speaking, keeping optimizer states in BF16 will degrade the final performance. Megatron-LM supports only FP16 and FP32 for optimizer states. (FP16 requires a separate scaling factor)

Here's a comparison between FP32/BF16 optimizer states on Megatron-LM:
image

**Comment by fegin** [comment]
@ebsmothers any thoughts on this?

**Comment by samsja** [comment]
> I don't think keeping optimizer states in BF16 is a good idea.
> 
> Generally speaking, keeping optimizer states in BF16 will degrade the final performance. Megatron-LM supports only FP16 and FP32 for optimizer states. (FP16 requires a separate scaling factor)
> 
> Here's a comparison between FP32/BF16 optimizer states on Megatron-LM: image

I also don't think that doing pure bf16 training makes sense even for sft. If the goal is to reduce memory footprint of the optimizer I think that adam8bit is a better tradeoff for low gpu count and with many gpu fsdp should make the optimizer state quite small on each gpu

**Comment by tianyu-l** [comment]
@joecummings sounds like we should revert this PR, as doing bf16 everywhere does not seem to be the right way to save memory. Wdyt?

**Comment by ebsmothers** [comment]
Sorry just getting caught up here. My two cents: pure bf16 should not preclude using optimizers like 8-bit Adam. In my mind it is still generally useful (and fairly standard: see e.g. Lightning’s true bf16 precision setting) to store model weights and gradients in bfloat16 without an extra higher-precision copy. Is 8-bit Adam currently supported by Titan? My impression was that it isn’t, but lmk if that’s mistaken.
